url65a69497a4_sweider_BankSystem_JavaLab8_tgz-pJ8-banksystem_lab4_LauncherJ8/src/banksystem/lab4/core/transaction/TransactionQueue.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+
+package banksystem.lab4.core.transaction;
+
+import banksystem.lab4.core.broker.IBroker;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+
+/**
+ *
+ * @author alex
+ */
+public class TransactionQueue implements Runnable{
+    private static final Logger logger = LogManager.getLogger(TransactionQueue.class);
+    
+    private static final int SLEEP_TIME = 500;
+    private final List<Transaction> mainTransactionsQueue;
+    private final List<Transaction> bufferTransactionQueue;
+    private final List<Transaction> executedTransactions;
+    private final IBroker broker;
+
+    public TransactionQueue(IBroker broker) {
+        List<Transaction> list = new ArrayList<>();
+        List<Transaction> list2 = new ArrayList<>();
+        this.mainTransactionsQueue = Collections.synchronizedList(list);
+        this.bufferTransactionQueue = Collections.synchronizedList(list2);
+        this.executedTransactions = new ArrayList<>();
+        this.broker = broker;
+    }
+    
+    public synchronized void add(Transaction transaction){
+        this.bufferTransactionQueue.add(transaction);
+        logger.info("Transaction added to buffer");
+    }
+    
+    private synchronized void updateFromBuffer(){
+        this.mainTransactionsQueue.addAll(this.bufferTransactionQueue);
+        this.bufferTransactionQueue.clear();
+    }
+    @Override
+    public void run() {        
+        while(true){
+            this.waitForTransactions();
+            this.processExistingTransactions();
+            this.removeExecutedTransactions();
+            this.updateFromBuffer();
+        }
+    }
+
+    private void waitForTransactions() {
+        while(this.mainTransactionsQueue.isEmpty()){
+            try {
+                Thread.sleep(TransactionQueue.SLEEP_TIME);
+                this.updateFromBuffer();
+            } catch (InterruptedException ex) { ex.printStackTrace(); }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)